examples/bond/scripts/run/bond-functions.js

```diff
@@ -0,0 +1,328 @@
+    fromAccount: masterAccount,
+    toAccountAddr: accountAddress,
+    amountMicroAlgos: 200e6,
+    payFlags: {}
+  };
+  await executeTransaction(deployer, algoTxnParams);
+}
+
+/**
+ * Creates DEX_i lsig, burn B_i tokens, issue B_i+1 tokens
+ * @param {Account} masterAccount
+ * @param {Account} creatorAccount
+ * @param {Account} storeManagerAccount
+ */
+async function createDex (deployer, masterAccount, creatorAccount, storeManagerAccount) {
+  asaInfo = deployer.getASAInfo('bond-token');
+  appInfo = deployer.getApp('bond-dapp-stateful.py', 'bond-dapp-clear.py');
+  let scInitParam = {
+    TMPL_APPLICATION_ID: appInfo.appID,
+    TMPL_OWNER: creatorAccount.addr,
+    TMPL_STORE_MANAGER: storeManagerAccount.addr
+  };
+  const issuerLsig = await deployer.loadLogic('issuer-lsig.py', scInitParam);
+
+  const deployTx = {
+    type: types.TransactionType.DeployASA,
+    sign: types.SignType.SecretKey,
+    fromAccount: creatorAccount,
+    asaName: 'new-bond-token',
+    payFlags: {}
+  };
+  // Create B_[i+1]
+  newAsaInfo = await executeTransaction(deployer, deployTx);
+  console.log(newAsaInfo);
+  const newIndex = newAsaInfo[assetID];
+
+  // Only store manager can allow opt-in to ASA for lsig
+  const optInTx = [
+    {
+      type: types.TransactionType.TransferAlgo,
+      sign: types.SignType.SecretKey,
+      fromAccount: storeManagerAccount,
+      toAccountAddr: issuerLsig.address(),
+      amountMicroAlgos: 0,
+      payFlags: {}
+    },
+    {
+      type: types.TransactionType.OptInASA,
+      sign: types.SignType.LogicSignature,
+      fromAccountAddr: issuerLsig.address(),
+      lsig: issuerLsig,
+      assetID: newIndex,
+      payFlags: {}
+    }
+  ];
+  await executeTransaction(deployer, optInTx);
+
+  scInitParam = {
+    TMPL_OLD_BOND: asaInfo.assetIndex,
+    TMPL_NEW_BOND: newIndex,
+    TMPL_APPLICATION_ID: appInfo.appID,
+    TMPL_STORE_MANAGER: storeManagerAccount.addr
+  };
+  const dexLsig = await deployer.loadLogic('dex-lsig.py', scInitParam);
+
+  await fundAccount(deployer, dexLsig.address());
+
+  optInTx[0].toAccountAddr = dexLsig.address();
+  optInTx[1].fromAccountAddr = dexLsig.address();
+  optInTx[1].lsig = dexLsig;
+  await executeTransaction(deployer, optInTx);
+
+  optInTx[1].assetID = asaInfo.assetIndex;
+  await executeTransaction(deployer, optInTx);
+
+  const globalState = await readGlobalStateSSC(deployer, storeManagerAccount.addr, appInfo.appID);
+  let total = 0;
+  for (const l of globalState) {
+    const key = Buffer.from(l.key, 'base64').toString();
+    if (key === 'total') {
+      total = l.value.uint;
+      break;
+    }
+  }
+
+  // Transfer total amount to dex lsig
+  const transferTx = {
+    type: types.TransactionType.TransferAsset,
+    sign: types.SignType.SecretKey,
+    fromAccount: creatorAccount,
+    toAccountAddr: dexLsig.address(),
+    amount: total,
+    assetID: newIndex,
+    payFlags: { totalFee: 1000 }
+  };
+  await executeTransaction(deployer, transferTx);
```

Comment:
this should be part of a group transaction -> see the spec. (and in fact you have it already in `groupTx`.

Reply:
done